### PR TITLE
TS-4994: make state configuration in ts_lua plugin

### PR DESCRIPTION
**Pull request by shukitchan**
@jpeach pls take a look when you have time.

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1181/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1074/ for details.

File: plugins/experimental/ts_lua/ts_lua.c [hunk 1]
```diff
@@ -64,18 +65,46 @@ TSRemapNewInstance(int argc, char *argv[], void **ih, char *errbuf, int errbuf_s
 {
   int fn;
   int ret;
+  int states                           = TS_LUA_MAX_STATE_COUNT;
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: Newline below here so clang format doesn't make it ugly ;)

File: plugins/experimental/ts_lua/ts_lua.c [hunk 2]
```diff
@@ -64,18 +65,46 @@ TSRemapNewInstance(int argc, char *argv[], void **ih, char *errbuf, int errbuf_s
...
+    }
 
-  if (argc < 3) {
+    if (opt == -1) {
+      break;
+    }
+  }
+
+  if (states > TS_LUA_MAX_STATE_COUNT || states < 1) {
+    snprintf(errbuf, errbuf_size, "[TSRemapNewInstance] - invalid state in option input");
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: Include the max and given values?

File: plugins/experimental/ts_lua/ts_lua.c [hunk 3]
```diff
@@ -391,12 +422,38 @@ TSPluginInit(int argc, const char *argv[])
     return;
   }
 
-  if (argc < 2) {
+  int states                           = TS_LUA_MAX_STATE_COUNT;
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: Newline.

File: plugins/experimental/ts_lua/ts_lua.c [hunk 4]
```diff
@@ -391,12 +422,38 @@ TSPluginInit(int argc, const char *argv[])
...
+      break;
+    }
+
+    if (opt == -1) {
+      break;
+    }
+  }
+
+  if (states > TS_LUA_MAX_STATE_COUNT || states < 1) {
+    TSError("[ts_lua][%s] invalid # of states from option input", __FUNCTION__);
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: Print the value and the max.

**Comment by zwoop** [comment]
Set the tags to the right please :).

**Comment by esmq2092** [comment]
change TS_LUA_MAX_STATE_COUNT from 256 to 64, still have the problem.

tail traffic.out 
PANIC: unprotected error in call to Lua API (not enough memory)
traffic_server: using root directory '/usr/local/trafficserver-7.1.0'
PANIC: unprotected error in call to Lua API (not enough memory)
traffic_server: received signal 11 (Segmentation fault)
traffic_server - STACK TRACE: 
traffic_server: using root directory '/usr/local/trafficserver-7.1.0'

**Comment by shukitchan** [comment]
1) you can keep lowering it and see if it helps.
2) you may be having too many scripts loaded or having too many variables holding too many data in memory at the same time. You can always check with - http://www.lua.org/manual/5.1/manual.html#pdf-collectgarbage , and see.

please do take note that luajit has a memory limitation - https://stackoverflow.com/questions/35155444/why-is-luajits-memory-limited-to-1-2-gb-on-64-bit-platforms

**Pull request merged**